test/e2e/fifo_queue.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
```

Comment:
I think not reimplementing the fronthalf of each daemon in the form of another controller x fifo x processing logic is good. Pop is usually a destructive operation though, so you could wrap the fifo in your type, and give it a Peek method. Just a suggestion, since I haven't done a deep review of this PR.